tests/test_ops.py

```diff
@@ -0,0 +1,277 @@
+    res = calc_tt_residual(
+        delays=tt_delay_seconds,
+        transponder_delays=transponders_delay,
+        twtt_model=twtt_model,
+    )
+    assert np.allclose(res, expected_residual)
+
+
+@pytest.mark.parametrize(
+    "transmit_vectors,reply_vectors,tt_delays,expected_a",
+    list(zip(TRANSMIT_VECTORS, REPLY_VECTORS, TT_DELAY_SECONDS, A_PARTIALS)),
+)
+def test_calc_partials(
+    transmit_vectors,
+    reply_vectors,
+    tt_delays,
+    expected_a,
+    num_transponders,
+    mean_sv,
+    b_cov,
+):
+    A_partials, B_cov, _, _ = calc_partials(
+        transmit_vectors=transmit_vectors,
+        reply_vectors=reply_vectors,
+        transponders_mean_sv=mean_sv,
+        delays=tt_delays,
+        num_transponders=num_transponders,
+    )
+    assert np.allclose(A_partials, expected_a)
+    assert np.array_equal(B_cov, b_cov)
+
+
+@pytest.mark.parametrize(
+    "gps_dict,transmit_vectors,expected_wm",
+    list(zip(GPS_DATASET[2:], TRANSMIT_VECTORS, WEIGHT_MATRIX)),
+)
+def test_calc_weight_matrix(
+    gps_dict, transmit_vectors, expected_wm, mean_sv, b_cov, travel_times_variance
+):
+    gps_dataseries = pd.Series(gps_dict)
+    # Add missing data
+    if "z" not in gps_dataseries:
+        gps_dataseries["z"] = 4511065.769
+    if "yy" not in gps_dataseries:
+        gps_dataseries["yy"] = 0.001007144325
+
+    gps_covariance_matrix = np.array(np.array_split(gps_dataseries[GPS_COV], 3)).astype(
+        "float64"
+    )
+    transmit_uv = np.array([calc_uv(v) for v in transmit_vectors])
+    weight_matrix = calc_weight_matrix(
+        transmit_uv=transmit_uv,
+        gps_covariance_matrix=gps_covariance_matrix,
+        transponders_mean_sv=mean_sv,
+        b_cov=b_cov,
+        travel_times_variance=travel_times_variance,
+    )
+    assert np.allclose(weight_matrix, expected_wm)
+
+
+@pytest.mark.parametrize(
+    "input_array,expected",
+    [
+        # 2-D case
+        (
+            np.array([[1, 1, 1, 0, 0, 0], [2, 2, 2, 0, 0, 0], [3, 4, 5, 0, 0, 0]]),
+            np.array([[1, 1, 1], [2, 2, 2], [3, 4, 5]]),
+        ),
+        # 1-D case
+        (np.array([1, 3, 4, 0, 7, 8, 0, 0, 0]), np.array([1, 3, 4, 0, 7, 8])),
+        # 3-D case, fails
+        (np.ones(shape=(2, 2, 3)), NotImplementedError),
+    ],
+)
+def test_clean_zeros(input_array, expected):
+    try:
+        result = clean_zeros(input_array=input_array)
+        assert np.array_equal(result, expected)
+    except Exception as e:
+        assert isinstance(e, expected)
+
+
+def test_calc_lsq_contrained():
+    X, XP, MX, MXP = calc_lsq_contrained(ATWA_SAMPLE, ATWF_SAMPLE)
+
+    # The results are in meters and we'd like to ensure cm
+    # precision, so 1e-4 should be very precise,
+    # might even be overkill
+    assert np.allclose(X, LSQ_RESULT[0], atol=1e-4)
+    assert np.allclose(XP, LSQ_RESULT[1], atol=1e-4)
+    assert np.allclose(MX, LSQ_RESULT[2], atol=1e-4)
+    assert np.allclose(MXP, LSQ_RESULT[3], atol=1e-4)
```

Comment:
@johnbdesanto Is this an okay precision tolerance to assume equivalents in the data? 1e-4?

Reply:
The legacy code reports positioning precise to 1e-4 m, with uncertainties that have more significant figures. Practically speaking, I think this is a good level although the code is probably going more precise under the hood. (From our discussion the other week, the criteria the old code uses is some small fraction of the standard deviation, which can be much smaller than 1e-4.) For reference, here is an excerpt for a sol_out file:

After iteration  3, rms residual =     45.85 cm, error factor =    10.255
 09 TRANSPONDER (23162 MEASUREMENTS):
   x =  -2690793.3252 +/- 1.46558D-04 m   del_e =      0.0466 +/- 1.50678D-04 m
   y =  -3835618.4714 +/- 1.48229D-04 m   del_n =     -0.1612 +/- 1.48716D-04 m
   z =   4310864.2626 +/- 1.49188D-04 m   del_v =         0.2385 +/- 1.44526D-04 m
   Lat. =  42.808288298 deg., Long. = -125.050730740 deg., Hgt.msl = -1499.1265 m
 10 TRANSPONDER (23162 MEASUREMENTS):
   x =  -2690858.9732 +/- 1.46558D-04 m   del_e =      0.0467 +/- 1.50679D-04 m
   y =  -3833685.5512 +/- 1.48229D-04 m   del_n =     -0.1613 +/- 1.48714D-04 m
   z =   4312494.0363 +/- 1.49188D-04 m   del_v =         0.2385 +/- 1.44527D-04 m
   Lat. =  42.828504308 deg., Long. = -125.064967519 deg., Hgt.msl = -1524.3355 m
 11 TRANSPONDER (23162 MEASUREMENTS):
   x =  -2689077.5316 +/- 1.46558D-04 m   del_e =      0.0465 +/- 1.50677D-04 m
   y =  -3834901.8001 +/- 1.48229D-04 m   del_n =     -0.1613 +/- 1.48715D-04 m
   z =   4312566.9845 +/- 1.49188D-04 m   del_v =         0.2385 +/- 1.44528D-04 m
   Lat. =  42.829154027 deg., Long. = -125.038583041 deg., Hgt.msl = -1494.8755 m